zcash_primitives/src/transaction/builder.rs

```diff
@@ -0,0 +1,615 @@
+                        },
+                        Note {
+                            g_d,
+                            pk_d,
+                            r: Fs::random(&mut self.rng),
+                            value: 0,
+                        },
+                    )
+                };
+
+                let esk = generate_esk();
+                let epk = dummy_note.g_d.mul(esk, &JUBJUB);
+
+                let (zkproof, cv) =
+                    prover.output_proof(&mut ctx, esk, dummy_to, dummy_note.r, dummy_note.value);
+
+                let cmu = dummy_note.cm(&JUBJUB);
+
+                let mut enc_ciphertext = [0u8; 580];
+                let mut out_ciphertext = [0u8; 80];
+                self.rng.fill_bytes(&mut enc_ciphertext[..]);
+                self.rng.fill_bytes(&mut out_ciphertext[..]);
+
+                OutputDescription {
+                    cv,
+                    cmu,
+                    ephemeral_key: epk.into(),
+                    enc_ciphertext,
+                    out_ciphertext,
+                    zkproof,
+                }
+            };
+
+            self.mtx.shielded_outputs.push(output_desc);
+        }
+
+        //
+        // Signatures
+        //
+
+        let mut sighash = [0u8; 32];
+        sighash.copy_from_slice(&signature_hash_data(
+            &self.mtx,
+            consensus_branch_id,
+            SIGHASH_ALL,
+            None,
+        ));
+
+        // Create Sapling spendAuth and binding signatures
+        for (i, (_, spend)) in spends.into_iter().enumerate() {
+            self.mtx.shielded_spends[i].spend_auth_sig = Some(spend_sig(
+                PrivateKey(spend.extsk.expsk.ask),
+                spend.alpha,
+                &sighash,
+                &JUBJUB,
+            ));
+        }
+        self.mtx.binding_sig = Some(
+            prover
+                .binding_sig(&mut ctx, self.mtx.value_balance.0, &sighash)
+                .map_err(|()| Error(ErrorKind::BindingSig))?,
+        );
+
+        Ok((
+            self.mtx.freeze().expect("Transaction should be complete"),
+            tx_metadata,
+        ))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use ff::{Field, PrimeField};
+    use rand::rngs::OsRng;
+    use sapling_crypto::jubjub::fs::Fs;
+
+    use super::{Builder, ErrorKind};
+    use crate::{
+        merkle_tree::{CommitmentTree, IncrementalWitness},
+        prover::mock::MockTxProver,
+        sapling::Node,
+        transaction::components::Amount,
+        zip32::{ExtendedFullViewingKey, ExtendedSpendingKey},
+        JUBJUB,
+    };
+
+    #[test]
+    fn fails_on_negative_output() {
+        let extsk = ExtendedSpendingKey::master(&[]);
+        let extfvk = ExtendedFullViewingKey::from(&extsk);
+        let ovk = extfvk.fvk.ovk;
+        let to = extfvk.default_address().unwrap().1;
+
+        let mut builder = Builder::new(0);
+        match builder.add_sapling_output(ovk, to, Amount(-1), None) {
```

Comment:
I think this pattern is more readable when testing functions that return `Result<..>`

Reply:
I agree it is more readable. I didn't do this here because I was matching the style used for testing the `Result` from `Builder::build()`, and that could not be done in the above style because it didn't build. I have now worked out that this is because neither `Transaction` nor `TransactionMetadata` implemented `PartialEq` (because performing a comparison on the whole `Result` requires both the `Ok` and `Err` types to implement both `Debug` and `PartialEq`). I will implement those and adjust the patterns.